File: frontend/test/services/declension_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:srbski_read/services/grammar_engine.dart';

/// Склонение существительных и прилагательных.
///
/// Примеры ЗДЕСЬ И В `server/internal/grammar/declension_test.go` одни и те же
/// намеренно: правила реализованы дважды, и разойтись они могут только молча.
/// Сайт и приложение обязаны называть одну и ту же форму — иначе человек,
/// который разбирает слово на телефоне и повторяет в браузере, увидит разное.
void main() {
  group('существительные', () {
    test('первая врста — защита от регресса', () {
      expect(GrammarEngine.matchNoun('kuća', 'Fem', 'kućom')?['Case'], 'Ins');
      expect(GrammarEngine.matchNoun('knjiga', 'Fem', 'knjizi')?['Case'], 'Dat');
    });

    test('женский род на согласный склоняется', () {
      // Раньше весь класс возвращал пустоту: noć, stvar, ljubav, kost.
      expect(GrammarEngine.matchNoun('noć', 'Fem', 'noći')?['Case'], 'Gen');
      expect(GrammarEngine.matchNoun('noć', 'Fem', 'noću')?['Case'], 'Ins');
      expect(GrammarEngine.matchNoun('noć', 'Fem', 'noćima')?['Number'], 'Plur');
      expect(
          GrammarEngine.matchNoun('ljubav', 'Fem', 'ljubavlju')?['Case'], 'Ins');
      expect(GrammarEngine.matchNoun('kost', 'Fem', 'košću')?['Case'], 'Ins');
      // После «р» йотования нет — только форма на -i.
      expect(GrammarEngine.matchNoun('stvar', 'Fem', 'stvari')?['Case'], 'Gen');
    });

    test('средний род с расширением основы', () {
      // ime → imena, а не «ima»; vreme → vremena, а не «vrema».
      expect(GrammarEngine.matchNoun('ime', 'Neut', 'imena')?['Case'], 'Gen');
      expect(GrammarEngine.matchNoun('ime', 'Neut', 'imenom')?['Case'], 'Ins');
      expect(
          GrammarEngine.matchNoun('vreme', 'Neut', 'vremena')?['Case'], 'Gen');
      expect(GrammarEngine.matchNoun('dete', 'Neut', 'deteta')?['Case'], 'Gen');
      // Средний род без расширения его не получает.
      expect(GrammarEngine.matchNoun('selo', 'Neut', 'sela')?['Case'], 'Gen');
      expect(GrammarEngine.matchNoun('polje', 'Neut', 'poljem')?['Case'], 'Ins');
    });

    test('беглое «а»', () {
      expect(GrammarEngine.matchNoun('otac', 'Masc', 'oca')?['Case'], 'Gen');
      expect(GrammarEngine.matchNoun('pas', 'Masc', 'psa')?['Case'], 'Gen');
      expect(GrammarEngine.matchNoun('momak', 'Masc', 'momka')?['Case'], 'Gen');
      expect(GrammarEngine.matchNoun('momak', 'Masc', 'momci')?['Number'],
          'Plur');
      expect(
          GrammarEngine.matchNoun('starac', 'Masc', 'starci')?['Number'],
          'Plur');
      // Родительный множественного — единственное место, где «а» возвращается.
      final momaka = GrammarEngine.matchNoun('momak', 'Masc', 'momaka');
      expect(momaka?['Case'], 'Gen');
      expect(momaka?['Number'], 'Plur');
      // Односложное «-ak» беглого «а» не имеет.
      expect(GrammarEngine.matchNoun('znak', 'Masc', 'znaka')?['Case'], 'Gen');
    });

    test('мужской род на -a склоняется по женскому типу', () {
      expect(GrammarEngine.matchNoun('tata', 'Masc', 'tate')?['Case'], 'Gen');
      expect(GrammarEngine.matchNoun('sudija', 'Masc', 'sudiju')?['Case'],
          'Acc');
    });

    test('множественное считает слоги, а не буквы', () {
      expect(GrammarEngine.matchNoun('grad', 'Masc', 'gradovi')?['Number'],
          'Plur');
      // По буквам «sport» получал «sporti».
      expect(GrammarEngine.matchNoun('sport', 'Masc', 'sportovi')?['Number'],
          'Plur');
      expect(GrammarEngine.matchNoun('muž', 'Masc', 'muževi')?['Number'],
          'Plur');
      expect(GrammarEngine.matchNoun('prozor', 'Masc', 'prozori')?['Number'],
          'Plur');
      // Слоговое «р» — слог: «vrt» односложное и берёт -ov-.
      expect(GrammarEngine.matchNoun('vrt', 'Masc', 'vrtovi')?['Number'],
          'Plur');
      // Исключения: по буквам получалось «danovi» и «zubovi».
      expect(GrammarEngine.matchNoun('dan', 'Masc', 'dani')?['Number'], 'Plur');
      expect(GrammarEngine.matchNoun('zub', 'Masc', 'zubi')?['Number'], 'Plur');
      expect(GrammarEngine.matchNoun('konj', 'Masc', 'konji')?['Number'],
          'Plur');
    });

    test('винительный одушевлённого опознаётся', () {
      // Строка для показа склеивает варианты через «/», поэтому опознание
      // обязано перебирать список форм, а не сравниваться со строкой.
      expect(GrammarEngine.matchNoun('čovek', 'Masc', 'čoveka'), isNotNull);
    });
  });

  group('прилагательные', () {
    test('вид различается в мужском и среднем роде', () {
      expect(
          GrammarEngine.adjectiveForm('dobar', 'Masc', 'Sing', 'Nom', false),
          'dobar');
      expect(GrammarEngine.adjectiveForm('dobar', 'Masc', 'Sing', 'Nom', true),
          'dobri');
      expect(GrammarEngine.adjectiveForm('dobar', 'Masc', 'Sing', 'Gen', false),
          'dobra');
      expect(GrammarEngine.adjectiveForm('dobar', 'Masc', 'Sing', 'Gen', true),
          'dobrog');
      expect(GrammarEngine.adjectiveForm('dobar', 'Neut', 'Sing', 'Gen', true),
          'dobrog');
    });

    test('женский род и множественное вид не различают', () {
      expect(GrammarEngine.adjectiveForm('dobar', 'Fem', 'Sing', 'Dat', false),
          GrammarEngine.adjectiveForm('dobar', 'Fem', 'Sing', 'Dat', true));
      expect(GrammarEngine.adjectiveForm('dobar', 'Masc', 'Plur', 'Gen', false),
          GrammarEngine.adjectiveForm('dobar', 'Masc', 'Plur', 'Gen', true));
      expect(GrammarEngine.adjectiveForm('dobar', 'Fem', 'Sing', 'Dat', false),
          'dobroj');
      expect(GrammarEngine.adjectiveForm('dobar', 'Masc', 'Plur', 'Gen', true),
          'dobrih');
    });

    test('мягкая основа меняет «о» на «е» — но не в женском роде', () {
      expect(GrammarEngine.adjectiveForm('vruć', 'Masc', 'Sing', 'Gen', true),
          'vrućeg');
      expect(GrammarEngine.adjectiveForm('vruć', 'Neut', 'Sing', 'Nom', false),
          'vruće');
      expect(GrammarEngine.adjectiveForm('vruć', 'Fem', 'Sing', 'Dat', false),
          'vrućoj');
    });

    test('беглое «а» в основе прилагательного', () {
      expect(GrammarEngine.adjectiveForm('hladan', 'Masc', 'Sing', 'Gen', true),
          'hladnog');
      expect(GrammarEngine.adjectiveForm('topao', 'Masc', 'Sing', 'Gen', true),
          'toplog');
      // Односложные беглого «а» не имеют.
      expect(GrammarEngine.adjectiveForm('star', 'Masc', 'Sing', 'Gen', true),
          'starog');
    });

    test('лемма в определённом виде разбирается', () {
      expect(GrammarEngine.adjectiveForm('veliki', 'Masc', 'Sing', 'Gen', true),
          'velikog');
      expect(
          GrammarEngine.adjectiveForm('veliki', 'Masc', 'Sing', 'Nom', false),
          'velik');
    });
  });

  group('разбор признаков', () {
    test('степень сравнения называется', () {
      final info = GrammarEngine.describe('ADJ', const {
        'Degree': 'Cmp',
        'Case': 'Nom',
        'Number': 'Sing',
        'Gender': 'Masc',
      });
      expect(info.facts.any((f) => f.label == 'Степень сравнения'), isTrue);
      expect(info.summary, contains('сравнительная степень'));
    });

    test('вид прилагательного называется и объясняется', () {
      final info = GrammarEngine.describe('ADJ', const {
        'Definite': 'Def',
        'Case': 'Gen',
        'Number': 'Sing',
        'Gender': 'Masc',
      });
      expect(info.facts.any((f) => f.label == 'Вид прилагательного'), isTrue);
      expect(info.why, contains('određeni'));
    });

    test('деепричастие больше не даёт пустую карточку', () {
      final info = GrammarEngine.describe('VERB', const {'VerbForm': 'Conv'});
      expect(info.facts.any((f) => f.label == 'Форма'), isTrue);
      expect(info.summary, isNotEmpty);
      expect(info.why, contains('деепричастие'));
    });

    test('трпни придев отличается от прилагательного', () {
      final info = GrammarEngine.describe('ADJ', const {
        'VerbForm': 'Part',
        'Voice': 'Pass',
        'Gender': 'Masc',
        'Number': 'Sing',
        'Case': 'Nom',
      });
      expect(info.facts.any((f) => f.label == 'Залог'), isTrue);
      expect(info.why, contains('трпни'));
    });

    test('в сводке род по-русски, а не сырым тегом', () {
      final info = GrammarEngine.describe('NOUN', const {
        'Case': 'Nom',
        'Number': 'Sing',
        'Gender': 'Masc',
      });
      expect(info.summary, contains('род: мужской'));
      expect(info.summary, isNot(contains('Masc')));
    });
  });

  group('спряжение', () {
    test('i-спряжение на -ati', () {
      expect(GrammarEngine.matchVerb('držati', 'držim')?['Person'], '1');
      expect(GrammarEngine.matchVerb('trčati', 'trči')?['Person'], '3');
    });

    test('a-спряжение осталось a-спряжением', () {
      expect(GrammarEngine.matchVerb('slušati', 'slušam')?['Person'], '1');
      expect(GrammarEngine.matchVerb('čitati', 'čitamo')?['Number'], 'Plur');
    });
  });
}

File: server/internal/grammar/declension_test.go
package grammar

import "testing"

// Склонение существительных проверяется примерами из первой сотни слов любого
// учебника: именно на них ошибка заметна человеку, который только учится.
func TestNounDeclension(t *testing.T) {
	cases := []struct {
		lemma, gender, number, caseKey, want string
	}{
		// Первая врста — она работала и раньше, поэтому здесь защита от регресса.
		{"kuća", "Fem", "Sing", "Gen", "kuće"},
		{"kuća", "Fem", "Sing", "Dat", "kući"},
		{"kuća", "Fem", "Sing", "Ins", "kućom"},
		{"knjiga", "Fem", "Sing", "Dat", "knjizi"},
		{"kuća", "Fem", "Plur", "Dat", "kućama"},

		// Третья врста: раньше весь класс возвращал пустоту.
		{"noć", "Fem", "Sing", "Gen", "noći"},
		{"noć", "Fem", "Sing", "Acc", "noć"},
		{"noć", "Fem", "Sing", "Ins", "noću / noći"},
		{"noć", "Fem", "Plur", "Dat", "noćima"},
		{"ljubav", "Fem", "Sing", "Ins", "ljubavlju / ljubavi"},
		{"kost", "Fem", "Sing", "Ins", "košću / kosti"},
		{"smrt", "Fem", "Sing", "Ins", "smrću / smrti"},
		// После «р» йотования нет — остаётся только форма на -i.
		{"stvar", "Fem", "Sing", "Ins", "stvari"},
		{"stvar", "Fem", "Sing", "Gen", "stvari"},

		// Мужской род на -a склоняется по женскому типу.
		{"tata", "Masc", "Sing", "Gen", "tate"},
		{"tata", "Masc", "Sing", "Ins", "tatom"},
		{"sudija", "Masc", "Sing", "Acc", "sudiju"},
		{"sudija", "Masc", "Plur", "Nom", "sudije"},

		// Средний род с расширением основы.
		{"ime", "Neut", "Sing", "Gen", "imena"},
		{"ime", "Neut", "Sing", "Ins", "imenom"},
		{"ime", "Neut", "Plur", "Nom", "imena"},
		{"ime", "Neut", "Plur", "Dat", "imenima"},
		{"vreme", "Neut", "Sing", "Gen", "vremena"},
		{"dete", "Neut", "Sing", "Gen", "deteta"},
		{"dugme", "Neut", "Sing", "Gen", "dugmeta"},
		// Средний род без расширения не должен его получить.
		{"selo", "Neut", "Sing", "Gen", "sela"},
		{"selo", "Neut", "Plur", "Nom", "sela"},
		{"polje", "Neut", "Sing", "Ins", "poljem"},

		// Беглое «а».
		{"otac", "Masc", "Sing", "Gen", "oca"},
		{"otac", "Masc", "Sing", "Voc", "oče"},
		{"otac", "Masc", "Sing", "Ins", "ocem"},
		{"pas", "Masc", "Sing", "Gen", "psa"},
		{"momak", "Masc", "Sing", "Gen", "momka"},
		{"momak", "Masc", "Sing", "Voc", "momče"},
		{"momak", "Masc", "Plur", "Nom", "momci"},
		{"starac", "Masc", "Sing", "Gen", "starca"},
		{"starac", "Masc", "Sing", "Voc", "starče"},
		{"starac", "Masc", "Plur", "Nom", "starci"},
		// Родительный множественного — единственное место, где «а» возвращается.
		{"momak", "Masc", "Plur", "Gen", "momaka"},
		{"starac", "Masc", "Plur", "Gen", "staraca"},
		// Односложное «-ak» беглого «а» не имеет.
		{"znak", "Masc", "Sing", "Gen", "znaka"},

		// Множественное мужского рода: слоги, а не буквы.
		{"grad", "Masc", "Plur", "Nom", "gradovi"},
		{"sport", "Masc", "Plur", "Nom", "sportovi"},
		{"front", "Masc", "Plur", "Nom", "frontovi"},
		{"muž", "Masc", "Plur", "Nom", "muževi"},
		{"ključ", "Masc", "Plur", "Nom", "ključevi"},
		{"prozor", "Masc", "Plur", "Nom", "prozori"},
		{"student", "Masc", "Plur", "Nom", "studenti"},
		// Слоговое «р» — слог: «vrt» односложное и берёт -ov-.
		{"vrt", "Masc", "Plur", "Nom", "vrtovi"},
		// Исключения из правила «односложное → длинное множественное».
		{"dan", "Masc", "Plur", "Nom", "dani"},
		{"zub", "Masc", "Plur", "Nom", "zubi"},
		{"gost", "Masc", "Plur", "Nom", "gosti"},
		{"konj", "Masc", "Plur", "Nom", "konji"},
		{"prst", "Masc", "Plur", "Nom", "prsti"},
		{"vuk", "Masc", "Plur", "Nom", "vuci"},

		// Остальной мужской род — защита от регресса.
		{"čovek", "Masc", "Sing", "Voc", "čoveče"},
		{"grad", "Masc", "Sing", "Ins", "gradom"},
		{"konj", "Masc", "Sing", "Ins", "konjem"},
		{"konj", "Masc", "Sing", "Voc", "konju"},
		{"grad", "Masc", "Sing", "Acc", "grad / grada"},
		{"čovek", "Masc", "Plur", "Nom", "ljudi"},
	}

	for _, c := range cases {
		got := declension(c.lemma, c.gender, c.number, c.caseKey)
		if got != c.want {
			t.Errorf("%s (%s, %s, %s) = %q, ожидалось %q",
				c.lemma, c.gender, c.number, c.caseKey, got, c.want)
		}
	}
}

// Ни одна клетка таблицы не должна остаться пустой у слов, которые движок
// объявляет разобранными: прочерк там читается как «программа не знает».
func TestDeclensionFillsEveryCell(t *testing.T) {
	words := []struct{ lemma, gender string }{
		{"kuća", "Fem"}, {"noć", "Fem"}, {"stvar", "Fem"}, {"ljubav", "Fem"},
		{"grad", "Masc"}, {"otac", "Masc"}, {"momak", "Masc"}, {"konj", "Masc"},
		{"tata", "Masc"}, {"selo", "Neut"}, {"ime", "Neut"}, {"polje", "Neut"},
	}
	for _, w := range words {
		for _, number := range []string{"Sing", "Plur"} {
			for _, caseKey := range CaseOrder {
				if declension(w.lemma, w.gender, number, caseKey) == "" {
					t.Errorf("%s (%s): пустая клетка %s %s",
						w.lemma, w.gender, number, caseKey)
				}
			}
		}
	}
}

// Опознание словоформы обязано работать на тех же примерах: правило строит
// парадигму от кандидата и принимает разбор, только если форма в ней нашлась.
func TestMatchNoun(t *testing.T) {
	cases := []struct {
		lemma, gender, form, wantCase, wantNumber string
	}{
		{"kuća", "Fem", "kućom", "Ins", "Sing"},
		{"noć", "Fem", "noću", "Ins", "Sing"},
		{"noć", "Fem", "noćima", "Dat", "Plur"},
		{"ime", "Neut", "imena", "Gen", "Sing"},
		{"otac", "Masc", "oca", "Gen", "Sing"},
		{"momak", "Masc", "momci", "Nom", "Plur"},
		{"grad", "Masc", "gradovima", "Dat", "Plur"},
		// «čoveka» — это и родительный, и винительный одушевлённого. Побеждает
		// родительный: он идёт раньше в CaseOrder и встречается чаще.
		{"čovek", "Masc", "čoveka", "Gen", "Sing"},
	}
	for _, c := range cases {
		feats, ok := MatchNoun(c.lemma, c.gender, c.form)
		if !ok {
			t.Errorf("%s: форма %q не опознана", c.lemma, c.form)
			continue
		}
		if feats["Case"] != c.wantCase || feats["Number"] != c.wantNumber {
			t.Errorf("%s %q: разобрано как %s %s, ожидалось %s %s",
				c.lemma, c.form, feats["Case"], feats["Number"], c.wantCase, c.wantNumber)
		}
	}
}

// Строка для показа склеивает варианты через «/», поэтому опознание обязано
// работать со списком форм. Иначе винительный одушевлённого не нашёлся бы
// никогда — сравнение шло бы с «čovek / čoveka».
func TestAccusativeKeepsBothFormsSeparately(t *testing.T) {
	forms := declensionForms("čovek", "Masc", "Sing", "Acc")
	if len(forms) != 2 || forms[0] != "čovek" || forms[1] != "čoveka" {
		t.Fatalf("формы винительного: %q", forms)
	}
	if got := declension("čovek", "Masc", "Sing", "Acc"); got != "čovek / čoveka" {
		t.Fatalf("для показа получилось %q", got)
	}
}

func TestSyllables(t *testing.T) {
	cases := map[string]int{
		"grad": 1, "sport": 1, "prozor": 2, "student": 2,
		"prst": 1, "vrt": 1, "trg": 1, "momak": 2, "starac": 2,
	}
	for word, want := range cases {
		if got := syllables(word); got != want {
			t.Errorf("syllables(%q) = %d, ожидалось %d", word, got, want)
		}
	}
}
